v2/elasticsearch-common/src/main/java/com/google/cloud/teleport/v2/elasticsearch/utils/ElasticsearchIO.java

```diff
@@ -1452,7 +1455,9 @@ public void processElement(ProcessContext context) throws Exception {
                     "{ \"update\" : %s }%n{ \"doc\" : %s, \"doc_as_upsert\" : true }%n",
                     documentMetadata, document));
           } else {
-            batch.add(String.format("{ \"create\" : %s }%n%s%n", documentMetadata, document));
+            // switch from create to index to allow upsert (beneficial in case of duplication due to
+            // timeout/retry)
+            batch.add(String.format("{ \"index\" : %s }%n%s%n", documentMetadata, document));
```

Comment:
Not sure about this change. This may be a breaking change. I image some template users may want it to fail if the target document already exists (e.g. they only move incremental changes to Elastic, not expecting any duplicates, and if there are any that's a configuration problem).
Maybe add a parameter to keep "create" as the default and switch to "index" only if the parameter is set?

Reply:
sure, I can do that.. though in the original ElasticsearchIO (in the beam repo) it uses index rather than create

Reply:
see: https://github.com/apache/beam/blob/master/sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java#L1576

Out of curiosity, why does this repo define its own ElasticsearchIO which is quite different?

Reply:
I think the only reason this repo has its own version is that at the time the template was created ElasticsearchIO wasn't released as part of Beam yet.
So `ElasticsearchIO` was simply copy-pasted from there.

Unfortunately now these two versions diverged a lot, so we need to do something about it, but I don't have a specific plan for now :) Ideally, we need to merge all changes from both repos into the Beam's repo, and make the template re-use `ElasticseachIO` from Beam SDK.

Reply:
Agree with @ggprod that using `index` is safer, considering that we may not have exactly-once guarantees.

But also +1 on @an2x on not carrying a forked version. I glanced over ElasticsearchIO here, and apparently, it doesn't seem that we would need to merge anything upstream, but it needs some further verification/tests.

Digging further, https://github.com/GoogleCloudPlatform/DataflowTemplates/commit/302d261e7256b448c47eb7daf093903d0a46acf1 has some explanations and mentions that it was a temporary solution:

```
/**
 * This class was included and changed as a temporary solution to cover Elasticsearch's requirement
 * of op_type for data streams. The required feature will be available with Apache Beam 2.33.0.
 * After migrating to Apache Beam 2.33.0 this class should be removed.
 * Jira ticket: <a href="https://issues.apache.org/jira/browse/BEAM-12601">Support append-only indices in ES output</a>.
 * Github pull request: <a href="https://github.com/apache/beam/pull/15257">[BEAM-12601] Add append-only option</a>.
 */
```